keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,421 @@
+```
+When the `shouldDelete` checks succeed, a delete volume request is initiated on the driver. This ensures that the volume
+is deleted in backend. 
+
+Once the volume is deleted from the backend, the finalizer can be removed. This allows the pv to be removed from the api
+server.
+
+```go
+// deleteVolumeOperation attempts to delete the volume backing the given
+// volume. Returns error, which indicates whether deletion should be retried
+// (requeue the volume) or not
+func (ctrl *ProvisionController) deleteVolumeOperation(ctx context.Context, volume *v1.PersistentVolume) error {
+	operation := fmt.Sprintf("delete %q", volume.Name)
+	klog.Info(logOperation(operation, "started"))
+	//....
+	//....
+	//....
+    klog.Infof("deleteVolumeOperation succeeded for the PV: %s", volume.Name)
+    klog.Info(logOperation(operation, "volume deleted"))
+    klog.Infof("deleteVolumeOperation attempting to remove the pv deletion protection finalizer for the PV: %s", volume.Name)
+    if ctrl.checkFinalizer(volume, pvDeletionProtectionFinalizer) {
+        // If there's a finalizer remove it to allow the pv deletion from the api server.
+        if err = ctrl.removePvDeletionProtectionFinalizer(volume); err != nil {
+            klog.Info(logOperation(operation, "failed to remove persistentvolume deletion protection finalizer: %v", err))
+            return err
+        }
+    }
+    //...
+    //...
+// Remaining code below skipped for brevity...
+// Please refer sig-storage-lib-external-provisioner/controller/controller.go for the full code.
+```
+
+### Test Plan
+
+#### E2E tests
+An e2e test to exercise deletion of PV prior to deletion of PVC for a `Bound` PV-PVC pair.
+
+
+### Graduation Criteria
+
+#### Alpha
+* Feedback
+* Fix of the issue behind a feature flag.
+* Unit tests and e2e tests outlined in design proposal implemented.
+
+#### Alpha -> Beta
+* Allow the Alpha fix to soak for one release.
+
+#### Beta -> GA
+* Deployed in production and have gone through at least one K8s upgrade.
+
+### Upgrade / Downgrade Strategy
+* Upgrade from old Kubernetes(1.22) to new Kubernetes(1.23) with `HonorPvReclaim` flag enabled
+
+In the above scenario, the upgrade will cause change in default behavior as described in the current KEP. Additionally,
+if there are PVs that have a valid associated PVC and deletion timestamp set, then a finalizer is added to the PV.
+
+* Downgrade from new Kubernetes(1.22) to old Kubernetes(1.21).
+
+In the above scenario, downgrade will retain the current behavior.
+
+### Version Skew Strategy
+The fix is part of `kube-controller-manager` of `external-provisioner`.
+
+1. `kube-controller-manager` is upgraded, but `external-provisioner` is not:
+
+In this case the changed behavior is observed only for in-tree plugins. The drivers would still have the issue. Users
+will observe the new annotation on the PV, but it does not have any effect since the `external-provisioner` is not
+updated
```

Comment:
In the new approach, PV controller will add a finalizer that the external-provisioner will not remove. Such PV will be in the API server forever.

It would make sense if PV controller adds the finalizer only to PV it is able to delete (i.e. in-tree ones, when CSI migration is disabled). The external provisioner would add finalizers only to PVs that it can delete too (i.e. the CSI driver matches). That would prevent weird issues when external-provisioner and Kubernetes versions do not match.

Reply:
Sure.

Reply:
Moved the checking/adding the new finalizer to syncVolume() in pv_controller.go.